// node-se-API-connector/lib/auth.js
var request = require("request");
var Q = require("q");
var utils = require('./utils');

/**
 * Creates a new APIkey with the given options.
 * 
 * ```javascript
 * options: {
 *      email: "String", 
 *      password: "String", 
 *      name: "String (Name of the created APIkey)",
 *      type: "int", 
 *      validUntil: "Date", 
 *      maxUses: "int"
 * }
 * ```
 *  
 * @param {Object} options Options-Object
 * @return {Promise}
 */
function key(options) {
    var deferred = Q.defer();

    if (typeof options === 'object') {
        request({
            url: utils.getAPIUrl("/auth/key"),
            method: 'POST',
            json: {
                email: options.email,
                password: options.password,
                name: options.name,
                type: options.type,
                validUntil: options.validUntil,
                maxUses: options.maxUses
            }
        }, function (err, res, body) {
            utils.apiCB(deferred, err, res, body);
        });
    } else {
        deferred.reject("No correct options-object given");
    }

    return deferred.promise;
}

(function ($) {
    $.login = "loginFKT";
    $.logout = "logoutFKT";
    $.key = key;
})(exports);
