// node-se-API-connector/lib/classes.js
var os = require("os");
var utils = require('./utils');

function RsmData(key, version, state, data){
    this.key = key;
    if(utils.isNumber(version)){
        this.version = version | 0;
    } else {
        this.version = 1;
    }
    this.state = state;
    this.data = data;
};

function NotifyMessage(guid, error, message, rsm){
    this.guid = guid;
    if(utils.isBoolean(error)){
        this.error = error;
    } else {
        this.error = true;
    }
    this.message = message;
    if(utils.isObject(rsm)){
        this.rsm = rsm;
    }
    this.utcMs = new Date().getTime();
}

function Container(customerId, name, realName, type){
    if(utils.isString(customerId))
        this.customerId = customerId;
    if(utils.isString(customerId))
        this.name = name;
    if(utils.isString(customerId))
        this.realName = realName;
    if(utils.isNumber(type)){
        this.type = type | 0;
    }else{
        this.type = 4;
    }
    
    this.system = {
        //TODO: get public IP from os.interfaces()
        ip: null,
        isServer: false,
        isVm: false,
        numberOfProcessors: os.cpus().length,
        totalRam: os.totalmem(),
        architecture: os.arch()
    };
    
    var lastStart = new Date();
    lastStart.setSeconds(lastStart.getSeconds() - (os.uptime() | 0));
    
    this.additional = {
        name: os.hostname(),
        version: os.release(),
        servicepack:"",
        lastStart: lastStart.getTime()
    };
    
    this.maxHeartbeatTimeout= 20;
}

(function($){
    $.RsmData = RsmData;
    $.NotifyMessage = NotifyMessage;
    $.Container = Container;
})(exports);
