// node-se-API-connector/lib/container.js
var request = require("request");
var Q = require("q");
var utils = require('./utils');

/**
 * Find a container by id.
 * 
 * @param {String} containerId
 * @return {Promise}
 */
function get(containerId) {
    var deferred = Q.defer();
    request({
        url: utils.getAPIUrl("/container"),
        qs: {
            id: containerId,
            apiKey: utils.getAPIkey()
        },
        method: 'GET',
        json: true
    }, function (err, res, body) {
        utils.apiCB(deferred, err, res, body);
    });

    return deferred.promise;
}

/**
 * Creates a new container.
 * 
 * @params {Object} options
 * @return {Promise}
 */
function create(options) {
    var deferred = Q.defer();

    request({
        url: utils.getAPIUrl("/container"),
        method: 'POST',
        json: {
            apiKey: utils.getAPIkey(),
            customerId: options.customerId,
            parentId: options.parentId,
            name: options.name,
            realName: options.realName,
            type: options.type,
            createNotify: options.createNotify,
            maxHeartbeatTimeout: options.maxHeartbeatTimeout,
            action: options.action,
            system: options.system,
            additional: options.additional
        }
    }, function (err, res, body) {
        utils.apiCB(deferred, err, res, body);
    });

    return deferred.promise;
}

(function ($) {
    $.get = get;
    $.create = create;
    $.update = "updateFKT";
    $.getAll = "getAllFKT";
    $.install = "installFKT";
    $.inventory = "inventoryFKT";
})(exports);
